oeps/architectural-decisions/oep-XXXX-modular-micro-frontend-domains.rst

```diff
@@ -0,0 +1,209 @@
+========================================
+OEP-XXXX: Modular Micro-frontend Domains
+========================================
+
+.. list-table::
+   :widths: 25 75
+
+   * - OEP
+     - :doc:`OEP-XXXX <oep-XXXX-modular-micro-frontend-domains>`
+   * - Title
+     - Modular Micro-frontend Domains
+   * - Last Modified
+     - 2022-11-09
+   * - Authors
+     - Adolfo R. Brandes <adolfo@tcril.org>
+   * - Arbiter
+     - Adam Stankiewicz <astankiewicz@2u.com>
+   * - Status
+     - Draft
+   * - Type
+     - Architecture
+   * - Created
+     - 2022-11-09
+   * - Review Period
+     - TBD
+   * - Resolution
+     - TBD
+   * - References
+     - TBD
+
+
+***********
+Terminology
+***********
+
+* *MFE*: a micro-frontend
+* *SPA*: single-page application
+* *DDD*: domain-driven development
+
+
+********
+Abstract
+********
+
+The decision to gradually implement a micro-frontend architecture for Open edX was a largely successful attempt at avoiding some limitations of the monolithic ``edx-platform`` codebase.  In particular, it allowed the project to integrate more and better features by leveraging newer frontend technology, all much faster than would've been possible otherwise.  At the same time, because features encapsulated in MFEs are very loosely coupled to each other, this architecture eliminates most integration bottlenecks, allowing individual teams working on any given feature to be a lot more *productive*.
+
+However, roughly 3 years into this gradual re-platforming, several drawbacks to the choice of complete MFE independence have been identified.  This OEP proposes a middle-ground based on DDD and other industry best-practices, where a minor increase in integration costs is traded for large gains in user experience, customizability, ease of deployment, and development efficiency.
+
+
+**********
+Motivation
```

Comment:
I think we need to be careful to zero in on what problems we hope MFE modularity (is composability the right word?) will solve.  There are a lot of problems listed here as motivation that are legitimate critiques of the architecture, but which I don't believe will be appreciably improved by this effort.

My take is that the main things we hope to get out of this are:

1. Reduced overall bundle size via shared files at runtime
2. Improved page load performance and cross-page UX via loading a 'shell' application once that contains both common UI elements, like headers and footers, and shared files.
3. Prioritizing "consistency" over "flexibility" - this is a choice we made early on in favor of flexibility.  We allowed all MFEs to have shared libraries, yes, but to use whatever version of those they wanted.  I think we probably went too far, and are learning we care more about consistency.

These correspond roughly to "Changes that affect multiple MFEs are costly" and "Inefficient use of network bandwidth" below.

Things this does not significantly impact:

1. Code reusability - Having a place to put shared code so we can reuse it - we have shared libraries that all MFEs use - Paragon, frontend-build, frontend-platform, the header and footer libraries - if you want to add a plugin interface, or consistent UI elements, we already have places to put those.
2. Documentation - loading our code differently doesn't impact the documentation written for our independent repos.
4. UX design - our design teams coordinate through Paragon to make consistent experiences.  A technical solution for code loading won't affect that much.  Prioritizing consistency of version numbers _will_ help, yes, but it won't change how designers work on its own.
5. Deprecation of old UIs - I'm not sure how this changes as a result of this.
6. Changes across MFEs that aren't covered by the shared libraries - I'm not convinced this is common... if something is used by many MFEs, it should be in a shared library already (i.e, paragon, frontend-platform, etc.)

Long story short, I'm wondering if we can tighten up this list of motivations so we can prove we're addressing them with the proposed new technical capabilities.